metric_learn/_util.py

```diff
@@ -12,39 +14,264 @@ def vector_norm(X):
+    Make sure that the 2D array has some minimum number of features
+    (columns). The default value of 1 rejects empty datasets.
+    This check is only enforced when X has effectively 2 dimensions or
+    is originally 1D and ``ensure_2d`` is True. Setting to 0 disables
+    this check.
+
+  warn_on_dtype : boolean (default=False)
+    Raise DataConversionWarning if the dtype of the input data structure
+    does not match the requested dtype, causing a memory copy.
+
+  estimator : str or estimator instance (default=None)
+    If passed, include the name of the estimator in warning messages.
 
   Returns
   -------
-  tuples_valid : object
-    The validated input.
+  X : `numpy.ndarray`
+    The checked input data array.
+
+  y: `numpy.ndarray` (optional)
+    The checked input labels array.
   """
-  # If input is scalar raise error
-  if np.isscalar(tuples):
-    raise ValueError(
-      "Expected 3D array, got scalar instead. Cannot apply this function on "
-      "scalars.")
-  # If input is 1D raise error
-  if len(tuples.shape) == 1:
-    raise ValueError(
-      "Expected 3D array, got 1D array instead:\ntuples={}.\n"
-      "Reshape your data using tuples.reshape(1, -1, 1) if it contains a "
-      "single tuple and the points in the tuple have a single "
-      "feature.".format(tuples))
-  # If input is 2D raise error
-  if len(tuples.shape) == 2:
-    raise ValueError(
-      "Expected 3D array, got 2D array instead:\ntuples={}.\n"
-      "Reshape your data either using tuples.reshape(-1, {}, 1) if "
-      "your data has a single feature or tuples.reshape(1, {}, -1) "
-      "if it contains a single tuple.".format(tuples, tuples.shape[1],
-                                              tuples.shape[0]))
+
+  context = make_context(estimator)
+
+  args_for_sk_checks = dict(accept_sparse=accept_sparse,
+                            dtype=dtype, order=order,
+                            copy=copy, force_all_finite=force_all_finite,
+                            ensure_min_samples=ensure_min_samples,
+                            ensure_min_features=ensure_min_features,
+                            warn_on_dtype=warn_on_dtype, estimator=estimator)
+  if y is None:
+    input = check_array(input, ensure_2d=False, allow_nd=True,
+                        copy=False, force_all_finite=False,
+                        accept_sparse=True, dtype=None,
+                        ensure_min_features=0, ensure_min_samples=0)
+  else:
+    input, y = check_X_y(input, y, ensure_2d=False, allow_nd=True,
+                         copy=False, force_all_finite=False,
+                         accept_sparse=True, dtype=None,
+                         ensure_min_features=0, ensure_min_samples=0,
+                         multi_output=multi_output,
+                         y_numeric=y_numeric)
+    # we try to allow the more possible stuff here
+  preprocessor_has_been_applied = False
+
+  if type_of_inputs == 'classic':

```

Comment:
This function is pretty long. Maybe we could break up the `type_of_inputs == 'classic'` and `'tuples'` cases into helper functions? They could return a boolean indicating if the preprocessor was applied, which would also help avoid the manual toggling here.